### PR TITLE
Sockend rooms with message as wrapper

### PR DESCRIPTION
**Pull request by orrgal1**
__rooms attribute in message to specify which rooms to send to

### CODE REVIEW

File: src/components/sockend.js [hunk 1]
```diff
@@ -88,27 +88,31 @@ module.exports = class Sockend extends Configurable(Component) {
...
-            subscriber.on('**', function(data) {
+            subscriber.on('**', function (data) {
                 if (this.event == 'cote:added' || this.event == 'cote:removed') return;
 
                 let topic = this.event.split('::');
                 let namespace = '';
-
-                if (topic.length > 1) {
-                    namespace += '/' + topic[0];
-                    topic = topic.slice(1);
```
Comment (dashersw) [inline src/components/sockend.js]: why did you remove this? do namespaces work with this?

File: src/components/sockend.js [hunk 2]
```diff
@@ -88,27 +88,31 @@ module.exports = class Sockend extends Configurable(Component) {
...
-                }
-
                 topic = topic.join('');
 
-                io.of(namespace).emit(topic, data);
+                let emitter = io.of(namespace);
+                if (data.__rooms) {
+                    const rooms = data.__rooms;
+                    delete data.__rooms;
+                    rooms.map((room) => {
```
Comment (dashersw) [inline src/components/sockend.js]: this is actually a forEach operation rather than a map.

File: README.md [hunk 3]
```diff
@@ -699,7 +699,21 @@ randomSubscriber.on('randomUpdate', (req) => {
 level with WebSockets over socket.io. `Sockend` makes `Responder`s and
 `Publisher`s available to the front-end and adhere to socket.io namespaces.
 It's the magic and the lost link for microservices. Without any configuration,
-you can expose APIs directly to the front-end.
+you can expose APIs directly to the front-end. 
+
+`Sockend` supports socket.io rooms. All you need to do is add `__rooms` attribute to
```
Comment (dashersw) [inline README.md]: `_rooms` means that it will always be an array. Isn't this a limiting factor? I imagine quite a big portion of the time it will be a single room. how about we accept both arrays and a single room name?

File: src/components/sockend.js (hunk 1 above)
Comment (orrgal1) [reply src/components/sockend.js]: might be a mistake, good catch let me check

File: README.md (hunk 3 above)
Comment (orrgal1) [reply README.md]: no problem. in that case we can support both __room and __rooms for clarity. but i kinda followed the same line as broadcasts:[], subscribesTo:[], etc..

Comment (dashersw) [reply README.md]: nice catch! except that broadcasts is 99% of the time an array.

for me the question is, should we have two separate values `__room` and `__rooms` or just have `__room` but it also accepts an array?

Comment (orrgal1) [reply README.md]: i think that would be less intuitive...
imho:
__room: 'room1' - good (as a convience)
__rooms:['room1','room2'] - good (can also be used as __rooms:['room1'] of course)
__rooms: 'room1' - not so good, not intuitive

what do you say?

Comment (dashersw) [reply README.md]: you are right. let's go with `__room` and `__rooms` then.

Comment (orrgal1) [reply README.md]: ok. on it. plus other fixes you requested...

File: src/components/sockend.js [hunk 4]
```diff
@@ -91,24 +91,36 @@ module.exports = class Sockend extends Configurable(Component) {
...
                     topic = topic.slice(1);
                 }
-
                 topic = topic.join('');
 
-                io.of(namespace).emit(topic, data);
+                let emitter = io.of(namespace);
+                if (data.__room) {
+                    data.__rooms = new Set(data.__rooms || []);
+                    data.__rooms.add(data.__room);
```
Comment (dashersw) [inline src/components/sockend.js]: do you maybe need to delete `data.__room` here?

File: test/sockend.js [hunk 5]
```diff
@@ -67,6 +67,32 @@ test.cb('Sockend simple pub&sub', (t) => {
...
+
+        client.on('published message', (msg) => {
+            t.deepEqual(msg, { content: 'simple content' });
+            t.end();
+        });
+
+        server.on('connection', (sock) => {
+            sock.join('room1');
+            publisher.sock.sock.on('connect', (sdf) => {
+                publisher.publish('published message', { content: 'simple content', __rooms: ['room1'] });
```
Comment (dashersw) [inline test/sockend.js]: we need tests for the scenario of multiple rooms with `__rooms` and also a single room with `__room`.

File: README.md [hunk 6]
```diff
@@ -710,6 +710,10 @@ Example:
 var socket = io.connect();
 var socketNamespaced = io.connect('/rnd');
 
+socket.on('randomUpdate',function(data){
```
Comment (dashersw) [inline README.md]: there are lint issues throughout the readme and the examples.

File: test/sockend.js (hunk 5 above)
Comment (orrgal1) [reply test/sockend.js]: yeah already on it...

File: src/components/sockend.js (hunk 4 above)
Comment (orrgal1) [reply src/components/sockend.js]: si senior. will do

File: README.md (hunk 6 above)
Comment (orrgal1) [reply README.md]: i'll run the lint fixer on them

Comment (orrgal1) [reply README.md]: how do you identify lint issues in readme?

Comment (orrgal1) [reply README.md]: you mean actually with your eyes?? or run eslint on it somehow?

Comment (dashersw) [reply README.md]: yes, actually with my eyes.

Comment (orrgal1) [reply README.md]: ok :) i'll try, not  sure i know all the lint rules exactly. are we  talking about spaces in the right places etc..?

Comment (dashersw) [reply README.md]: :) yes. but as an alternative, you can c/p the examples into a source file and run the linter.

Comment (orrgal1) [reply README.md]: on it

Comment (orrgal1) [reply README.md]: ok i've gone over entire readme. fixed a few lint issues...

File: examples/sockend/sockend.js [hunk 7]
```diff
@@ -1,8 +1,17 @@
...
+try {
+    cote = require('cote');
+} catch (e) {
+    cote = require('../../');
+}
+let app = require('http').createServer(handler),
     io = require('socket.io').listen(app),
     fs = require('fs');
 
+io.on('connection', (socket)=>{
```
Comment (dashersw) [inline examples/sockend/sockend.js]: Missing spaces here as well.

File: README.md [hunk 8]
```diff
@@ -753,6 +757,10 @@ const cote = require('cote'),
     io = require('socket.io').listen(app),
     fs = require('fs');
 
+io.on('connection', (socket)=>{
```
[the remark this answers is not in the document]
Comment (orrgal1) [reply README.md]: i run the linter on it but it doesnt change. could we be using different rules? aren't those defined in package.json?

File: examples/sockend/sockend.js (hunk 7 above)
Comment (orrgal1) [reply examples/sockend/sockend.js]: same here...

File: README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (orrgal1) [reply README.md]: not sure what spaces are missing...

File: examples/sockend/sockend.js (hunk 7 above)
Comment (dashersw) [reply examples/sockend/sockend.js]: it should be `io.on('connection', (socket) => {`

File: README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (dashersw) [reply README.md]: no — idk what the problem is with the linter, but this should be `io.on('connection', (socket) => {`

[the remark this answers is not in the document]
Comment (orrgal1) [reply README.md]: ok will fix

**Comment by dashersw** [comment]
@orrgal1 thank you very much!
